README
# MultiBoilerplate extension for MediaWiki

The MultiBoilerplate extension allows a boilerplate to be selected from a drop
down box located above the edit form. By default this shows only on creation of
new pages. When loading a boilerplate, it will completely replace whatever text
is already in the edit form.

## Dependencies
Version 2.0.0 and later require MediaWiki >= 1.24.

## Installation
See the regular installation instructions for MediaWiki extensions:
	https://www.mediawiki.org/wiki/Manual:Extensions#Installing_an_extension

## Configuration

### Main configuration
The main configuration is of the available boilerplates; this is done *either*:
	
* through `$wgMultiBoilerplateOptions` in `LocalSettings.php`, by filling the
  $wgMultiBoilerplateOptions array with a list of boilerplate names that
  correspond to templates, like so:  
  ```$wgMultiBoilerplateOptions[ "My Boilerplate" ] = "Template:My Boilerplate";```
* Or through system message `MediaWiki:Multiboilerplate`, which uses the following
  format:  
  `My Boilerplate|Template:My Boilerplate`  
  You can also create headers to separate boilerplates, for example:
 	
        == Pretty Templates ==
        * My Boilerplate|Template:My Boilerplate
        == Ugly Templates ==
        * Their Boilerplate|Template:Their Boilerplate


### Additional configuration options

* `$wgMultiBoilerplateDiplaySpecialPage`: false by default. if set to true,
  will add to the wiki a page named `Special:MultiBoilerplates` that shows
  the currently configured boilerplates.

* `$wgMultiBoilerplateOverwrite`: false by default. If true, shows the
  boilerplates dropdown even on pre-existing page. The selected boilerplate 
  will completed overwrite the current contents.

## Change log
See files RELEASE-NOTES and HISTORY for a complete change log.

## Credits
Originally by Robert Leverington (minuteelectron), with additional contributions
by Al Maghi and Dror S [FFS].
See the [commit log](https://git.wikimedia.org/log/mediawiki%2Fextensions%2FMultiBoilerplate)
for a full list of contributers.





HISTORY
Change notes from older releases. For current info see RELEASE-NOTES.

### 1.8.0, 2009-07-31
Optional special page

### 1.7, 2009-04-08
Get boilerplates from content message when used, instead of user language message.


### 1.6, 2008-02-26
* Added README.
* `ltrim()` now used instead of `preg_replace()`.

### 1.5, 2008-02-06
Now handles `<noinclude>` and `<includeonly>` tags.

### 1.4, 2008-02-06
* It is now (optionally) possible to specify boilerplates in a MediaWiki namespace message
  instead of a `LocalSettings.php]` configuration variable.
* The extension now uses `descriptionmsg` (allowing description internationalisation).
* Part of the code that used direct database interfacing to check for article existence has been
  replaced with a call to the `$wgTitle` global.
* The `$wgMultiBoilerplateThings` global has been renamed to `$wgMultiBoilerplateOptions`
  for a greater level of self-explanation.
* The boilerplate selection box is no longer displayed if there are no options.

### 1.3, 2008-01-19
* Use `Xml::` class methods instead of manually created HTML
* Convert all double-quotes to single-quotes,
* Add an optional message `multiboilerplate-label` to display a label before
  the drop-down box
* Rename `multiboilerplate-select` to `multiboilerplate-legend` and
  `multiboilerplate-load` to `multiboilerplate-submit` to be less ambiguous.

### 1.2, 2008-01-07
Refactor code, fix major bug that made the extension useless and add comments.

### 1.1, 2007-12-18
Original version.


MultiBoilerplate.php
<?php

/**
 * The MultiBoilerplate extension allows a boilerplate to be selected from a drop
 * down box located above the edit form. By default this shows only on creation of
 * new pages. When loading a boilerplate, it will completely replace whatever text
 * is already in the edit form.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 * http://www.gnu.org/copyleft/gpl.html
 *
 * @file
 * @ingroup Extensions
 *
 * @link https://www.mediawiki.org/wiki/Extension:MultiBoilerplate
 * @author Robert Leverington <robert@rhl.me.uk>
 * @author Al Maghi
 * @author Dror S. [FFS] <FreedomFighterSparrow@gmail.com
 * @copyright Copyright © 2007 - 2009 Robert Leverington.
 * @copyright Copyright © 2009 Al Maghi.
 * @license http://www.gnu.org/copyleft/gpl.html GNU General Public License 2.0 or later
 * @version 2.0.0
 *
 * @TODO use the core ?preload=boilerplate when $wgMultiBoilerplateOverwrite == false
 * @TODO ajax-load the boilerplate, if possible
 * @TODO per-namespace boilerplates
 */

$extensionCredits = array(
	'path'           => __FILE__,
	'name'           => 'MultiBoilerplate',
	'version'        => '2.0.0',
	'license-name'   => 'GPL-2.0+',
	'url'            => 'https://www.mediawiki.org/wiki/Extension:MultiBoilerplate',
	'author'         => array(
		'Robert Leverington',
		'Al Maghi',
		'Dror S. [FFS]'
	)
);

$GLOBALS[ 'wgExtensionCredits' ][ 'other' ][] = $extensionCredits + array(
	'descriptionmsg' => 'multiboilerplate-desc',
);
$GLOBALS[ 'wgExtensionCredits' ][ 'specialpage' ][] = $extensionCredits + array(
	'descriptionmsg' => 'specialboilerplate-desc',
);

// Default configuration variables.
/**
 * Array of boilerplate names to boilerplate pages to load, for example:
 * $wgMultiBoilerplateOptions[ 'My Boilerplate' ] = 'Template:My Boilerplate';
 * If set to false then the MediaWiki:multiboilerplate message is used to configure
 * boilerplates in the format of:
 * "* Boilerplate Name|Template:Boilerplate Template"
 */
$GLOBALS['wgMultiBoilerplateOptions'] = array();
// Whether or not to show the form when editing pre-existing pages.
$GLOBALS['wgMultiBoilerplateOverwrite'] = false;

// To display a special page listing defined boilerplates, set *before* require_once:
// $wgMultiBoilerplateDiplaySpecialPage = true;
$dir = __DIR__;
$GLOBALS[ 'wgAutoloadClasses' ][ 'MultiBoilerplateHooks' ] = "$dir/MultiBoilerplate.hooks.php";
$GLOBALS[ 'wgMessagesDirs']['MultiBoilerplate'] = "$dir/i18n";
$GLOBALS[ 'wgExtensionMessagesFiles' ][ 'MultiBoilerplate' ] = "$dir/MultiBoilerplate.i18n.php";

if ( isset( $GLOBALS[ 'wgMultiBoilerplateDiplaySpecialPage' ] )
	&& $GLOBALS[ 'wgMultiBoilerplateDiplaySpecialPage' ] === true
) {
	$GLOBALS[ 'wgAutoloadClasses' ][ 'SpecialBoilerplates' ] = "$dir/SpecialBoilerplates.php";
	$GLOBALS[ 'wgExtensionMessagesFiles' ]['MultiBoilerplateAlias'] =
		"$dir/MultiBoilerplate.alias.php";
	$GLOBALS[ 'wgSpecialPages' ][ 'Boilerplates' ] = 'SpecialBoilerplates';
	$GLOBALS[ 'wgSpecialPageGroups' ][ 'Boilerplates' ] = 'pages';
}

$GLOBALS['wgHooks'][ 'EditPage::showEditForm:initial' ][] =
	'MultiBoilerplateHooks::onEditPageShowEditFormInitial';


SpecialBoilerplates.php
<?php
/**
 * Special:boilerplates, provides a list of MediaWiki:Multiboilerplate or $wgMultiBoilerplateOptions
 * For more info see http://mediawiki.org/wiki/Extension:Multiboilerplate
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 * http://www.gnu.org/copyleft/gpl.html
 *
 * @file
 * @ingroup Extensions
 * @ingroup SpecialPage
 * @author Al Maghi
 *
 * @TODO fix code duplication (here and in MultiBoilerplate.hooks.php)
 * @TODO Use special page to actually edit [[MediaWiki:MultiBoilerplate]]?
 */

class SpecialBoilerplates extends IncludableSpecialPage {

	function __construct() {
		parent::__construct( 'Boilerplates' );
		$this->mIncludable = true;
	}

	public function execute( $par ) {
		global $wgMultiBoilerplateOptions;
		$output = $this->getOutput();

		// No options found in either configuration file, abort.
		if ( !isset( $wgMultiBoilerplateOptions ) ) {
			return true;
		}

		if ( !$this->mIncluding ) {
			$this->setHeaders();
			$output->addWikiMsg( 'multiboilerplate-special-pagetext' );
		}
		if ( is_array( $wgMultiBoilerplateOptions ) ) {
			if ( !$this->mIncluding ) {
				$output->addWikiMsg( 'multiboilerplate-special-define-in-localsettings' );
			}
			foreach ( $wgMultiBoilerplateOptions as $name => $template ) {
				$output->addWikiText( "* [[$template]]\n" );
			}
		} else {
			if ( !$this->mIncluding ) {
				$output->addWikiMsg( 'multiboilerplate-special-define-in-interface' );
			}
			$things = explode( "\n", str_replace( "\r", "\n", str_replace(
				"\r\n", "\n", wfMessage( 'multiboilerplate' )
			) ) ); // Ensure line-endings are \n
			foreach ( $things as $row ) {
				if ( substr( ltrim( $row ), 0, 1 ) === '*' ) {
					$row = ltrim( $row, '* ' ); // Remove astersk & spacing from start of line.
					$row = explode( '|', $row );
					if ( !isset( $row[ 1 ] ) ) {
						return true; // Invalid syntax, abort
					}
					$output->addWikiText( "* [[$row[1]|$row[0]]]\n" );
				}
			}
		}

		return true;
	}
}



MultiBoilerplate.hooks.php
<?php
/**
 * Hooks for MultiBoilerplate extension
 *
 * * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 * http://www.gnu.org/copyleft/gpl.html
 *
 * @file
 * @ingroup Extensions
 *
 * @TODO de-duplicate code between this and SpecialBoilerplates
 */

class MultiBoilerplateHooks {


	/**
	 * EditPage::showEditForm:initial hook
	 *
	 * Generate a boilerplate selection input on top of the edit page
	 *
	 * @param EditPage $editPage the current EditPage object.
	 * @param OutputPage $out object.
	 * @return true
	 */
	public static function onEditPageShowEditFormInitial( EditPage $editPage, OutputPage $out ) {
		global $wgParser, $wgMultiBoilerplateOptions, $wgMultiBoilerplateOverwrite;
		$title = $out->getTitle();
		$request = $out->getRequest();

		// If $wgMultiBoilerplateOverwrite is true then detect whether
		// the current page exists or not and if it does return true
		// to end execution of this function.
		if ( !$wgMultiBoilerplateOverwrite && $title->exists() ) {
			return true;
		}

		// Generate the options list used inside the boilerplate selection box.
		// If $wgMultiBoilerplateOptions is an array then use that, else fall back
		// to the MediaWiki:Multiboilerplate message.
		if ( is_array( $wgMultiBoilerplateOptions ) ) {
			$options = '';
			foreach ( $wgMultiBoilerplateOptions as $name => $template ) {
				$selected = false;
				if ( $request->getVal( 'boilerplate' ) === $template ) {
					$selected = true;
				}
				$options .= Xml::option( $name, $template, $selected );
			}
		} else {
			$things = wfMessage( 'multiboilerplate' )->inContentLanguage()->text();
			$options = '';
			$headingFound = 0;
			$things = explode( "\n", str_replace(
				"\r", "\n", str_replace( "\r\n", "\n", $things )
			) ); // Ensure line-endings are \n
			foreach ( $things as $row ) {
				if ( preg_match( '/==\s*(.*)\s*==/', $row, $optGroupText ) ) {
					if ( $headingFound ) {
						$options .= '</optgroup>';
					}
					$headingFound = true;
					$options.='<optgroup label="' . htmlspecialchars( $optGroupText[1] ) . '">';
				} elseif ( substr( ltrim( $row ), 0, 1 ) === '*' ) {
					$row = ltrim( $row, '* ' ); // Remove astersk & spacing from start of line.
					$rowParts = explode( '|', $row );
					if ( !isset( $rowParts[ 1 ] ) ) {
						return true; // Invalid syntax, abort
					}
					$selected = false;
					if ( $request->getVal( 'boilerplate' ) === $rowParts[ 1 ] ) {
						$selected = true;
					}
					$options .= Xml::option( $rowParts[ 0 ], $rowParts[ 1 ], $selected );
				}

			}
			if ( $headingFound ) {
				$options .= '</optgroup>';
			}
		}

		// No options found in either configuration file, abort.
		if ( $options == '' ) {
			return true;
		}

		// Append the selection form to the top of the edit page.
		$editPage->editFormPageTop .=
			Xml::openElement( 'form', array(
				'id' => 'multiboilerplateform',
				'name' => 'multiboilerplateform',
				'method' => 'get',
				'action' => $title->getEditURL() )
			)
			. Xml::openElement( 'fieldset' )
			. Xml::element( 'legend', null, wfMessage( 'multiboilerplate-legend' ) )
			. Xml::openElement( 'label' )
			. wfMessage( 'multiboilerplate-label' )->text()
			. Xml::openElement( 'select', array( 'name' => 'boilerplate' ) )
			. $options
			. Xml::closeElement( 'select' )
			. Xml::closeElement( 'label' )
			. ' '
			. Html::Hidden( 'action', 'edit' )
			. Html::Hidden( 'title', $request->getText( 'title' ) )
			. Xml::submitButton( wfMessage( 'multiboilerplate-submit' ) )
			. Xml::closeElement( 'fieldset' )
			. Xml::closeElement( 'form' );

		// If the Load button has been pushed replace the article text with the boilerplate.
		if ( $request->getText( 'boilerplate', false ) ) {
			$boilerplateTitle = Title::newFromText( $request->getVal( 'boilerplate' ) );
			if ( !$boilerplateTitle->exists() ) {
				$out->addHTML(
					'<strong class="error">'
					. $out->msg( 'multiboilerplate-nonexistant-page' )->params(
						$request->getText( 'boilerplate')
					)->text()
					. '</strong>'
				);
			} else {
				$boilerplate   = new WikiPage( $boilerplateTitle );
				$parser        = $wgParser->getFreshParser();  // Since MW 1.24
				$parserOptions = is_null( $parser->getOptions() ) ? new ParserOptions : $parser->getOptions();
				$content       = $parser->getPreloadText(
					$boilerplate->getContent()->getWikitextForTransclusion(),
					$boilerplateTitle,
					$parserOptions
				);

				$editPage->textbox1 = $content;
			}
		}

		return true;
	}

}
